### PR TITLE
chore(ui): migrate InputGroup to Typescript

### PR DESCRIPTION
**Pull request by guoda-puidokaite**
# Summary

- Migrated the `InputGroup` component to `TypeScript`, enhancing code quality, maintainability, and scalability.
- Utilised static typing to minimize runtime errors and improve the development experience.
- Refactored the component for cleaner, more maintainable code.

# Changes Made

**TypeScript Migration**:

- Converted `InputGroup ` component files to `TypeScript`.
- Added comprehensive type annotations.
- Introduced a deprecated `InputGroup` component for `Javascript` files.

**Refactoring**:

- Made necessary adjustments to address TypeScript-detected issues.
- Further refactored for cleaner and more understandable code.

**Testing**:

- Added edge cases:
  - renders without children
  - renders with a mix of enabled and disabled children
  - allows custom attributes to be passed to children
- Removed `NativeSelect` component from tests and stories, as not yet converted. 
- Organised tests by what they check for better readability and maintenance.
- Verified component behaviour using `Storybook`.
- Updated tests to use `vitest`.

**Documentation:**

- Updated `Storybook` documentation.

Screenshot: Deprecated `.js` code on the left, migrated `.ts` code on the right

Screenshot 2024-10-28 at 17 23 36

**Release:**

- Generated a `changeset` file.

# Related Issues

- https://github.com/cloudoperators/juno/issues/231

# Testing Instructions

1. Run `Storybook`
2. Run `InputGroup ` tests

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have made corresponding changes to the documentation (if applicable).
- [x] My changes generate no new warnings or errors.

### CODE REVIEW

**Review by barsukov — COMMENTED** [review]
Leave it as a comment, from my side I did not see any problems, so you could easily convert it :) good job 🥇

**Comment by barsukov** [comment]
Ups, tests, lintl and typescript has a diferent opinion haha :)

**Comment by guoda-puidokaite** [comment]
Thank You for initial review though! Still a draft, working on the tests and story, issue in PR description. 🤪

File: packages/ui-components/src/components/InputGroup/InputGroup.component.tsx [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ */
+export const InputGroup: React.FC<InputGroupProps> = ({
+  children = null,
+  className = "",
+  variant = "default",
+  disabled = false,
+  ...props
+}) => {
+  // Clone each child element with inherited variant and disabled props
+  const modifiedChildren = React.Children.map(children, (child) =>
```
Comment (barsukov) [inline packages/ui-components/src/components/InputGroup/InputGroup.component.tsx:67]: Could you clarify this action if it is neccessery ?

Comment (edda) [reply packages/ui-components/src/components/InputGroup/InputGroup.component.tsx:67]: I believe it's necessary. Unless you know another way to apply props to arbitrary `children` that you haven't created yourself.

File: packages/ui-components/src/components/InputGroup/InputGroup.component.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+   */
+  variant?: VariantTypes
+  /**
+   * If true, all elements within the InputGroup will be disabled.
+   * Individual elements can override this setting if needed.
+   */
+  disabled?: boolean
+}
+
+type OverrideProps = { variant?: VariantTypes; disabled?: boolean }
```
[the remark this answers is not in the document]
Comment (edda) [reply packages/ui-components/src/components/InputGroup/InputGroup.component.tsx:33]: Maybe `InheritProps` would be fitting? It's about the children inheriting props from the parent

File: packages/ui-components/src/components/InputGroup/InputGroup.component.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ */
+const cloneElementWithInheritedProps = (
+  child: ReactNode,
+  parentVariant: VariantTypes,
+  parentDisabled: boolean
+): ReactNode => {
+  if (!isValidElement<OverrideProps>(child)) return child
+
+  const combinedProps = {
+    variant: child.props.variant ?? parentVariant,
```
Comment (edda) [inline packages/ui-components/src/components/InputGroup/InputGroup.component.tsx:49]: What's your reason for replacing the `||` operator in the original component with the nullish coalescing operator? I think `||` is more appropriate here as it reacts to all falsy values not only `null` and `undefined`

In the future (not now) we need to revisit this in general. I think it makes more sense for the parent's prop value to always overwrite the child if it exists. But this component is still WiP for a reason, it's not completely finished

File: packages/ui-components/src/components/InputGroup/InputGroup.component.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  ...props
+}) => {
+  // Clone each child element with inherited variant and disabled props
+  const modifiedChildren = React.Children.map(children, (child) =>
+    cloneElementWithInheritedProps(child, variant, disabled)
+  )
+
+  return (
+    <Stack
+      className={`juno-input-group juno-input-group-${variant} ${disabled ? "juno-input-group-disabled" : ""} ${className}`}
```
[the remark this answers is not in the document]
Comment (edda) [reply packages/ui-components/src/components/InputGroup/InputGroup.component.tsx:73]: Agreed. Makes it easier to read if the class name logic is extracted into its own method as we do in other components

**Review by edda — CHANGES_REQUESTED** [review]

File: packages/ui-components/src/deprecated_js/InputGroup/InputGroup.test.js [hunk 5]
```diff
@@ -5,11 +5,11 @@
 
 import * as React from "react"
 import { render, screen } from "@testing-library/react"
-import { InputGroup } from "./index"
+import { InputGroup } from "../../deprecated_js/InputGroup"
```
[the remark this answers is not in the document]
Comment (guoda-puidokaite) [reply packages/ui-components/src/deprecated_js/InputGroup/InputGroup.test.js:8]: Hmmm, I saw that all deprecated ones were keeping tests? Is that wrong?

File: packages/ui-components/src/deprecated_js/InputGroup/input-group.scss [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  
+  // prevent double-borders for identical buttons next to each other:
+  .juno-button-default + .juno-button-default,
+  .juno-button-primary + .juno-button-primary,
+  .juno-button-default-primary-danger + .juno-button-primary-danger,
+  .juno-button-subdued + .juno-button-subdued {
+    border-left: 0;
+  }
+
+}
```
[the remark this answers is not in the document]
Comment (guoda-puidokaite) [reply packages/ui-components/src/deprecated_js/InputGroup/input-group.scss:32]: It's the deprecated one so 🙈

File: packages/ui-components/src/components/InputGroup/InputGroup.stories.tsx [hunk 7]
```diff
@@ -4,11 +4,11 @@
  */
 
 import React from "react"
-import { InputGroup } from "./index.js"
-import { Button } from "../../deprecated_js/Button/index"
-import { NativeSelect } from "../NativeSelect/index"
-import { NativeSelectOption } from "../NativeSelectOption/index"
-import { TextInput } from "../../deprecated_js/TextInput/index"
+
+import { InputGroup } from "./"
```
[the remark this answers is not in the document]
Comment (guoda-puidokaite) [reply packages/ui-components/src/components/InputGroup/InputGroup.stories.tsx:8]: Could you expand on what you mean? 🙈

**Review by barsukov — DISMISSED** [review]
LGTM from my side good job 👍

File: packages/ui-components/src/components/InputGroup/InputGroup.component.tsx [hunk 8]
```diff
@@ -43,15 +43,24 @@ const cloneElementWithInheritedProps = (
...
   return React.cloneElement(child, combinedProps)
 }
 
+const getClassNames = (
+  className: string,
+  variant: VariantTypes,
+  disabled: boolean,
+  additionalClassName: string
+): string => {
+  return `${className} ${className}-${variant} ${disabled ? `${className}-disabled` : ""} ${additionalClassName}`
```
Comment (edda) [inline packages/ui-components/src/components/InputGroup/InputGroup.component.tsx:61]: One comment regarding using string interpolation to construct class names. This is fine in this case because those are custom class names but it wouldn't work for tailwind class names. So be careful with string interpolation in classnames. Find more info in the tailwind documentation about dynamic classnames: https://tailwindcss.com/docs/content-configuration#dynamic-class-names

Comment (edda) [reply packages/ui-components/src/components/InputGroup/InputGroup.component.tsx:61]: Two things here:
1. To stay consistent with other components the className function should only produce the classNames that deal with the styling of the element. The best practise is to specify the semantic name (`className` or `juno-input-group` here) in the jsx/tsx markup. Same with the `className` from props. That way it's easier to see when looking at the markup what the respective elements are and in case of the `className` from props for complex components there are multiple choices which inner element to apply those class names to and it's better if you can see this in the jsx/tsx markup section.
2. We typically write class name strings in multiple lines because this makes it easier to read.
```suggestion
  return `
      ${className}-${variant} 
      ${disabled ? `${className}-disabled` : ""} 
  `
```

File: packages/ui-components/src/components/InputGroup/InputGroup.component.tsx [hunk 9]
```diff
@@ -68,11 +77,10 @@ export const InputGroup: React.FC<InputGroupProps> = ({
...
   )
 
+  const inputGroupClassName = getClassNames("juno-input-group", variant, disabled, className)
+
   return (
-    <Stack
-      className={`juno-input-group juno-input-group-${variant} ${disabled ? "juno-input-group-disabled" : ""} ${className}`}
-      {...props}
-    >
+    <Stack className={inputGroupClassName} {...props}>
```
Comment (edda) [inline packages/ui-components/src/components/InputGroup/InputGroup.component.tsx:83]: See explanation above:
```suggestion
    <Stack className={`juno-input-group ${inputGroupClassName} ${className}`} {...props}>
```

**Review by edda — CHANGES_REQUESTED** [review]
For consistency!! :)

**Review by edda — APPROVED** [review]
🚀

**Review by barsukov — APPROVED** [review]
LGTM! good job
